recipe-client-addon/lib/StudyStorage.jsm

```diff
@@ -0,0 +1,108 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * @typedef {Object} Study
+ * @property {string} name
+ *   Unique name of the study
+ * @property {number} version
+ *   Study add-on version number
+ * @property {string} description
+ *   Description of the study and its intent.
+ * @property {string} studyStartDate
+ *   ISO-formatted date string of when the study was started.
+ * @property {string} studyEndDate
+ *   ISO-formatted date string of when the study was ended.
+ * @property {string} addonId
+ *   Add-on ID for this particular study.
+ * @property {boolean} active
+ *   Is the study still running?
+ */
+
+"use strict";
+
+const {utils: Cu} = Components;
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "IndexedDB", "resource://gre/modules/IndexedDB.jsm");
+
+this.EXPORTED_SYMBOLS = ["StudyStorage"];
+
+const DB_OPTIONS = {
+  namespace: "shield-study-addons",
+  version: 1,
+  storage: "persistent",
+  schema: {
+    name: { autoIncrement: false, unique: true },
+    version: { autoIncrement: false, unique: false },
+    description: { autoIncrement: false, unique: false },
+    studyStartDate: { autoIncrement: false, unique: false },
+    studyEndDate: { autoIncrement: false, unique: false },
+    addonId: { autoIncrement: false, unique: false },
+    active: { autoIncrement: false, unique: false },
+  },
+};
+
+const dbName = "shield-studies";
+let database;
+
+async function getDatabase() {
+  if (!database) {
+    database = await IndexedDB.open(dbName, DB_OPTIONS, db => {
+      db.createObjectStore(dbName, { keyPath: "name" });
+    });
+  }
+  return database;
+}
+
+this.StudyStorage = {
+  async clearAllData() {
+    if (database) {
+      await database.objectStore(dbName, "readwrite").clear();
+    }
+  },
+
+  async closeConnection() {
+    if (database) {
+      await database.close();
+      database = null;
+    }
+  },
+
+  async getStudy(studyName) {
+    const db = await getDatabase();
+
+    return db.objectStore(dbName, "readonly").get(studyName);
+  },
+
+  async saveStudy(study) {
```

Comment:
Yes, using it from testing wouldn't really work - we'd have to package it separately.

I'm not really up-to-scratch with schema checks (I mostly meant that we'd need checks *somewhere*, given that previously you could pass in just about anything), but it seems like ajv is 28k of JS once it's minified? Personally, that feels like overkill for what is needed here, but I'm not going to object to either going with ajv (or some other off-the-shelf solution) or hand-rolling some basic checks.